### PR TITLE
fix: LineNumber presentation for Git word-diff

### PR DESCRIPTION
**Pull request by gerhardol**
## Proposed changes

Tune the heuristics to get word-diff line numbers. If the diff only had added or removed chars, only added/removed line numbers were presented.

There are still many situations where the Git output is incomplete, the line numbers will be incorrect still.
Comments added to the code, trying to explain the issues.

## Screenshots 

File to compare is in a private branch, did not find a good example in recent commits

image

with #11868 

image

### Before

see line 20

image

### After

image

## Test methodology 

tests updated

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: src/app/GitUI/Editor/Diff/DiffLineNumAnalyzer.cs [hunk 1]
```diff
@@ -150,25 +152,60 @@ public static DiffLinesInfo Analyze(TextEditorControl textEditor, bool isCombine
...
-    private static bool IsPlusLineInCombinedDiff(string line)
-    {
-        return line.StartsWith("++") || line.StartsWith("+ ") || line.StartsWith(" +");
-    }
+            return textMarkers.Count == 1
 
-    private static bool IsMinusLineInCombinedDiff(string line)
-    {
-        return line.StartsWith("--") || line.StartsWith("- ") || line.StartsWith(" -");
+                    // start may be intended, if a new block of changes starts with white spaces
```
Comment (mstv) [inline src/app/GitUI/Editor/Diff/DiffLineNumAnalyzer.cs:168]:
```suggestion
                    // start may be indented, if a new block of changes starts with white spaces
```

File: src/app/GitUI/Editor/Diff/DiffLineNumAnalyzer.cs [hunk 2]
```diff
@@ -150,25 +152,60 @@ public static DiffLinesInfo Analyze(TextEditorControl textEditor, bool isCombine
...
+        {
+            // The expected marker color for a line type, for heuristics.
+
+            return lineType is (DiffLineType.Minus or DiffLineType.MinusLeft)
+                ? (AppSettings.ReverseGitColoring.Value
+                    ? textMarker.Color == AppColor.AnsiTerminalRedBackNormal.GetThemeColor()
+                    : textMarker.ForeColor == AppColor.AnsiTerminalRedForeNormal.GetThemeColor())
+                : (AppSettings.ReverseGitColoring.Value
+                    ? textMarker.Color == AppColor.AnsiTerminalGreenBackNormal.GetThemeColor()
+                    : textMarker.ForeColor == AppColor.AnsiTerminalGreenForeNormal.GetThemeColor());
```
Comment (mstv) [inline src/app/GitUI/Editor/Diff/DiffLineNumAnalyzer.cs:188]: Consider reading `AppSettings.ReverseGitColoring.Value` at most once per document.

**Review by mstv — COMMENTED** [review]
My first chosen example "23 22 SetText(ref text);" seems to be one of "the line numbers will be incorrect still." (not requesting a change)

Off topic:
The base commit also crashes when trying to display word diff of `tests/app/UnitTests/GitUI.Tests/Editor/Diff/SampleGitWord.diff`.

File: src/app/GitUI/Editor/Diff/DiffLineNumAnalyzer.cs (hunk 2 above)
Comment (gerhardol) [reply src/app/GitUI/Editor/Diff/DiffLineNumAnalyzer.cs:188]: What is the cost? This is used in other parts too.

**Comment by gerhardol** [comment]
> My first chosen example "23 22 SetText(ref text);" seems to be one of "the line numbers will be incorrect still." (not requesting a change)

That is the Git output, not much can be done.
Your diff is often more useful.

> The base commit also crashes when trying to display word diff

Noticed that too, have not investigated

File: src/app/GitUI/Editor/Diff/DiffLineNumAnalyzer.cs (hunk 2 above)
Comment (gerhardol) [reply src/app/GitUI/Editor/Diff/DiffLineNumAnalyzer.cs]: Similar could be done in the Difftastic highlighter, other seem OK

Comment (mstv) [reply src/app/GitUI/Editor/Diff/DiffLineNumAnalyzer.cs]: ~1000-4000 settings evaluations take 1 ms. This can add up.

Comment (vbjay) [reply src/app/GitUI/Editor/Diff/DiffLineNumAnalyzer.cs]: Cache at begining of process. Store colors in something like a dictionary of some key and the color value.  Just make sure dictionary is reloaded if settings change or even if new analyzer created.

**Pull request merged**